hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/function/json/SingleStoreJsonArrayInsertFunction.java

```diff
@@ -0,0 +1,114 @@
+
+	public SingleStoreJsonArrayInsertFunction(TypeConfiguration typeConfiguration) {
+		super( typeConfiguration );
+	}
+
+	@Override
+	public void render(
+			SqlAppender sqlAppender,
+			List<? extends SqlAstNode> arguments,
+			ReturnableType<?> returnType,
+			SqlAstTranslator<?> translator) {
+		final Expression json = (Expression) arguments.get( 0 );
+		final Expression jsonPath = (Expression) arguments.get( 1 );
+		final List<JsonPathHelper.JsonPathElement> jsonPathElements = JsonPathHelper.parseJsonPathElements( translator.getLiteralValue(
+				jsonPath ) );
+		final SqlAstNode value = arguments.get( 2 );
+		final int arrayIndex = getArrayIndex( jsonPathElements );
+		if ( jsonPathElements.size() > 1 ) {
+			sqlAppender.appendSql( "json_set_json(" );
+			json.accept( translator );
+			buildJsonPath( sqlAppender, jsonPath, jsonPathElements );
+			sqlAppender.appendSql( ", " );
+		}
+		sqlAppender.appendSql( "case when json_get_type(json_extract_json(" );
+		json.accept( translator );
+		buildJsonPath( sqlAppender, jsonPath, jsonPathElements );
+		sqlAppender.appendSql( ")) = 'array' THEN " );
+		buildJsonArrayInsertValue( sqlAppender, value );
+		sqlAppender.appendSql( "json_extract_json(" );
+		json.accept( translator );
+		buildJsonPath( sqlAppender, jsonPath, jsonPathElements );
+		sqlAppender.appendSql( "), " );
+		sqlAppender.appendSql( arrayIndex );
+		sqlAppender.appendSql( ", 0, " );
+		value.accept( translator );
+		sqlAppender.appendSql( ") ELSE " );
+		buildJsonArrayInsertValue( sqlAppender, value );
+		sqlAppender.appendSql( "json_build_array(json_extract_json(" );
+		json.accept( translator );
+		buildJsonPath( sqlAppender, jsonPath, jsonPathElements );
+		sqlAppender.appendSql( "))," );
+		sqlAppender.appendSql( arrayIndex );
+		sqlAppender.appendSql( ", 0, " );
+		value.accept( translator );
+		sqlAppender.appendSql( ") END" );
+		if ( jsonPathElements.size() > 1 ) {
+			sqlAppender.appendSql( ')' );
+		}
+	}
+
+	private static int getArrayIndex(List<JsonPathHelper.JsonPathElement> jsonPathElements) {
+		if ( jsonPathElements.isEmpty() ) {
+			throw new QueryException( "SingleStore json_array_insert function requires at least one json path element" );
+		}
+		JsonPathHelper.JsonPathElement lastPathElement = jsonPathElements.get( jsonPathElements.size() - 1 );
+		if ( !( lastPathElement instanceof JsonPathHelper.JsonIndexAccess ) ) {
+			throw new QueryException(
+					"SingleStore json_array_insert function last path parameter must be an array index element" );
+		}
+		return ( (JsonPathHelper.JsonIndexAccess) lastPathElement ).index();
+	}
+
+	private static boolean isNumeric(SqlAstNode value) {
+		return value instanceof UnparsedNumericLiteral<?>;
+	}
+
+	private static void buildJsonArrayInsertValue(SqlAppender sqlAppender, SqlAstNode value) {
+		sqlAppender.appendSql( "json_splice_" );
+		sqlAppender.appendSql( isNumeric( value ) ? "double(" : "string(" );
```

Comment:
Wouldn't it be easier and more correct if you used the `xxx_json` functions and pass the json value through using the `to_json()` function?